### DISCUSSION TITLE
Named url routes

### DISCUSSION THREAD

**Discussion by piotrpog** [craftcms/cms#8592] [Ideas]
When we create routes and then want to use them in templates, we have to enter route structure manually in template file.

This is not DRY aproach because if for some reason we decide to change route url structure, we also have to change every part of template that used this route.

I propose adding slug field for each route and adding some kind of route template tag that would take in route slug as well as all route tokens. Kinda like [url](https://docs.djangoproject.com/pl/2.1/ref/templates/builtins/#url) tag works in Django framework.

**Comment by brandonkelly** [comment]
Not sure I’m following. Can you elaborate on your idea a bit?

**Comment by piotrpog** [comment]
Sure. Lets say we create route for user profile page - `'author/<username>'` directing to specific template.

Next, we want to insert link to specific profile in templates - we have to create it like this:
`<a href="author/{{username}}">link name</a>`

Then if we change route to for example `'profile/<username>` in routes settings, old `<a href="author/{{username}}">` stays in template and link stops working.

If we had ability to give routes unique names, we could use them in templates using some kind of tag, like this:

`{% url 'profile-route' username %}`

Thats how it works in django.

**Comment by brandonkelly** [comment]
Thanks for clarifying!

**Comment by khalwat** [comment]
Related: https://github.com/craftcms/cms/issues/983

**Comment by AugustMiller** [comment]
Upvoted!

There appear to be [hints at this, within Yii](https://www.yiiframework.com/doc/guide/2.0/en/runtime-routing#creating-urls) (although, the routing in Craft is arguably much deeper than a basic CRUD app). Rails also has similar (but horribly opaque) “[path helpers](https://guides.rubyonrails.org/routing.html#path-and-url-helpers).”

**Comment by missmatsuko** [comment]
I think this would be useful for Craft Commerce as well. I followed the (Commerce v1) example where the location of the templates determined the location of the store, then the client changed their mind about where the store should be located. I had to change my folder structure and references to the shop location.

**Comment by AugustMiller** [comment]
@missmatsuko This might be a good use for Yii `@alias`es—many of these pages (i.e. `/checkout` or `/cart`) won't have a corresponding Entry/Element in Craft… except for Products (which do have a URI format).

That said, I can sympathize with the litany of changes required to migrate whole groups of pages like this… my best recommendation here is to hide _everything_ in your templates folder, and only expose some templates via `routes.php`, so at least your incoming routes can be changed in one place. Of course, this doesn't help the link generating component—but the `@alias` might help a bit, despite being more verbose, in some circumstances:

```twig
<a href="{{ url(alias('@store/checkout')) }}">Checkout</a>
```

**Comment by khalwat** [comment]
Right, I was going to suggest using aliases for this... seems like it solves the problem nicely.

**Comment by piotrpog** [comment]
This might be a little late, but i create workaround solution for named routes:
http://craftsnippets.com/articles/url-macro

**Comment by piotrpog** [comment]
If anyone is interested, I just released a plugin that adds such functionality:

https://plugins.craftcms.com/named-routes
